internal/models/refresh_token.go

```diff
@@ -41,6 +42,14 @@ type GrantParams struct {
 	FactorID *uuid.UUID
 
 	SessionNotAfter *time.Time
+
+	UserAgent string
+	IP        string
+}
+
+func (g *GrantParams) FromRequest(r *http.Request) {
```

Comment:
Maybe we can rename this function to be more descriptive like:
```suggestion
func (g *GrantParams) GetSessionDetails(r *http.Request) {
```

Reply:
To me `FillRequestDetails` is better I suppose as `Get` signals a getter and this is not a getter.